bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -91,87 +91,162 @@
-                    public Object addingService(ServiceReference reference) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
-                        return provider;
-                    }
+    /**
+     * Activates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     * @throws Exception
+     */
+    protected void activate(ComponentContext componentContext) throws Exception {
+        this.bundleContext = componentContext.getBundleContext();
+        this.ruleEngine = new RuleEngine(bundleContext, moduleTypeRegistry);
+        this.ruleEngine.setStatusInfoCallback(this);
+        super.activate(bundleContext);
+    }
 
-                    @Override
-                    public void modifiedService(ServiceReference reference, Object service) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
+    /**
+     * Deactivates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     */
+    protected void deactivate(ComponentContext componentContext) {
+        super.deactivate();
+        // removeManagedProvider(managedProvider);
+        ruleEngine.dispose();
+        ruleEngine = null;
+    }
 
-                    }
+    /**
+     * Bind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void setModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = moduleTypeRegistry;
+    }
 
-                    @Override
-                    public void removedService(ServiceReference reference, Object service) {
-                    }
+    /**
+     * Unbind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void unsetModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = null;
+    }
 
-                });
-        templateProviderTracker.open();
+    /**
+     * Bind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void setTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = templateRegistry;
+    }
 
+    /**
+     * Unbind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void unsetTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = null;
+    }
+
+    protected void setStorageService(StorageService storageService) {
+        // this.storageService = storageService;
```

Comment:
Kay, do you have some pattern for relation between other registries and their managed providers?
Are they 1:1? As I see in the ThinkRegistry it has 0:1 dynamic relation to ManagedThingProvider.
This means if ManagedThingProvider is disappeared the ThinkRegistry continues to work, but you can't add,remove and update thinks. Is this desired behavior in the rules? As I remember there was a reported problem that if you want to add rule you have to wait for two services RuleRegistry and Managed provider (which was too complex) and then start the add operation.
Isn't be easier to have 1:1 dependency to managed provider because the managed provider depends only to storage service and it is a system service which shouldn't disappeared frequently?